src/seat/input_method.rs

```diff
@@ -0,0 +1,391 @@
+                        WEnum::Value(cause) => cause,
+                        WEnum::Unknown(value) => {
+                            warn!(
+                                "Unknown `text_change_cause`: {}. Assuming not input method.",
+                                value
+                            );
+                            ChangeCause::Other
+                        }
+                    },
+                    ..imdata.pending_state.clone()
+                }
+            }
+            Event::ContentType { hint, purpose } => {
+                imdata.pending_state = InputMethodEventState {
+                    active: imdata.pending_state.active.with_content_type(),
+                    content_hint: match hint {
+                        WEnum::Value(hint) => hint,
+                        WEnum::Unknown(value) => {
+                            warn!(
+                                "Unknown content hints: 0b{:b}, ignoring.",
+                                ContentHint::from_bits_retain(value)
+                                    - ContentHint::from_bits_truncate(value)
+                            );
+                            ContentHint::from_bits_truncate(value)
+                        }
+                    },
+                    content_purpose: match purpose {
+                        WEnum::Value(v) => v,
+                        WEnum::Unknown(value) => {
+                            warn!("Unknown `content_purpose`: {}. Assuming `normal`.", value);
+                            ContentPurpose::Normal
+                        }
+                    },
+                    ..imdata.pending_state.clone()
+                }
+            }
+            Event::Done => {
+                imdata.pending_state = InputMethodEventState {
+                    active: imdata.pending_state.active.with_done(),
+                    ..imdata.pending_state.clone()
+                };
+                imdata.current_state = imdata.pending_state.clone();
+                imdata.serial += 1;
+                data.handle_done(qh, &input_method, &imdata.current_state)
+            }
+            Event::Unavailable => data.handle_unavailable(qh, &input_method),
+            _ => unreachable!(),
+        };
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    struct Handler {}
+
+    impl InputMethodHandler for Handler {
+        fn handle_done(
+            &self,
+            _qh: &QueueHandle<Self>,
+            _input_method: &ZwpInputMethodV2,
+            _state: &InputMethodEventState,
+        ) {
+        }
+
+        fn handle_unavailable(&self, _qh: &QueueHandle<Self>, _input_method: &ZwpInputMethodV2) {}
+    }
+
+    delegate_input_method!(Handler);
+
+    fn assert_is_manager_delegate<T>()
+        where T: wayland_client::Dispatch<crate::reexports::protocols_misc::zwp_input_method_v2::client::zwp_input_method_manager_v2::ZwpInputMethodManagerV2, crate::globals::GlobalData>,
```

Comment:
```suggestion
        where T: wayland_client::Dispatch<ZwpInputMethodManagerV2, crate::globals::GlobalData>,
```
It's already imported into the scope, we might as well use it.